File: logb-py/statement.py
import module
from entity import Entity, EntityType

import sys
__thismodule__ = sys.modules[__name__]

class StatementType(object):
    def __init__(self, name, mod):
        ':type mod: Module'
        self.name = name
        self.module = module
        mod.add_statement_type(self)

class Statement(Entity):
    def __init__(self, typ, arguments):
        ':type typ: EntityType'
        super(self.__class__, self).__init__(__thismodule__.et_statement)
        self.statement_type = typ
        self.structure = {}
        self.structure["arguments"] = arguments

    def deepcopy(self):
        rval = Statement(self.statement_type, {})
        rval.structure = self.copy_structure(self.structure)
        return rval

m_statement = module.Module("statement")
et_statement = EntityType("statement", m_statement)
et_statement.has_structure = True  # The structure are the statement parameters.

File: logb-java/src/org/logb/core/Statement.java
package org.logb.core;

public class Statement extends Entity {
	static public EntityType ENTITY_TYPE_STATEMENT;
	
	public static void initialize() {
		ENTITY_TYPE_STATEMENT = new EntityType("Statement", new Module("Core"), false, true);
	}
	
	public Statement(StatementType type) {
		this(type, null);
	}

	public Statement(StatementType type, EntityStructureBase arguments) {
		super(ENTITY_TYPE_STATEMENT);
		this.statementType = type;
		setStructure(arguments);
	}

	@Override
	public EntityStructureBase deepcopy() {
		EntityStructureBase structure = null;
		if(this.getStructure() != null) {
			structure = this.getStructure().deepcopy();
		}
		return new Statement(this.statementType, structure);
	}

	@Override
	public boolean equals(EntityStructureBase otherBase) {
		if (!(otherBase instanceof Statement)) {
			return false;
		}

		Statement other = (Statement) otherBase;
		return other.statementType.equals(this.statementType)
				&& other.getStructure().equals(this.getStructure());
	}
	
	@Override
	public String toString() {
		String rval = statementType.getName();
		if(getStructure() != null) {
			rval += " " + getStructure().toString();
		}
		return rval;
	}

	private final StatementType statementType;
}